src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Cohost/CohostProjectSnapshot.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.AspNetCore.Razor.ProjectEngineHost;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Telemetry;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Cohost;
+
+internal class CohostProjectSnapshot : IProjectSnapshot
+{
+    private static readonly EmptyProjectEngineFactory s_fallbackProjectEngineFactory = new();
+    private static readonly (IProjectEngineFactory Value, ICustomProjectEngineFactoryMetadata)[] s_projectEngineFactories = ProjectEngineFactories.Factories.Select(f => (f.Item1.Value, f.Item2)).ToArray();
+
+    private readonly Project _project;
+    private readonly DocumentSnapshotFactory _documentSnapshotFactory;
+    private readonly ITelemetryReporter _telemetryReporter;
+    private readonly ProjectKey _projectKey;
+    private readonly Lazy<RazorConfiguration> _lazyConfiguration;
+    private readonly Lazy<RazorProjectEngine> _lazyProjectEngine;
+    private readonly AsyncLazy<ImmutableArray<TagHelperDescriptor>> _tagHelpersLazy;
+    private readonly Lazy<ImmutableDictionary<string, ImmutableArray<string>>> _importsToRelatedDocumentsLazy;
+
+    public CohostProjectSnapshot(Project project, DocumentSnapshotFactory documentSnapshotFactory, ITelemetryReporter telemetryReporter, JoinableTaskContext joinableTaskContext)
+    {
+        _project = project;
+        _documentSnapshotFactory = documentSnapshotFactory;
+        _telemetryReporter = telemetryReporter;
+        _projectKey = ProjectKey.From(_project)!.Value;
```

Comment:
💡 `AssumeNotNull()` works for nullable structs too. It'll throw if null and return the value.